lerna build

evgeniuz opened the pull request:
Fixes #107

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ConsenSys/ethql?pullRequest=117) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ConsenSys/ethql?pullRequest=117) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ConsenSys/ethql?pullRequest=117) it.</sub>

evgeniuz commented:
Added *very* rough lerna build, still lots to do, but packages are split and building. Instructions are:
```
yarn install
lerna bootstrap
lerna run build:ts
node package/ethql/dist/index.js
```
Things still to do (non-complete list):

- [ ] move tests into packages and ensure they're working
- [ ] move package lifecycle scripts into packages
- [ ] move dependencies into packages
- [ ] make global lifecycle scripts working (like building docker image)
- [ ] document all changes made to actual source code: I've had couple of issues, mostly related to TypeScript types and imports, which are resolved now, but I wan't someone to check if these are ok.

evgeniuz commented:
Implemented lerna build in the best way possible (I hope :smiley:):

- It uses new TypeScript feature (build mode and project references), this allows to run complete build and watching across all packages at once (also replaced `nodemon` with `tsc-watch`, as it's much simpler wrapper around `tsc` and more suitable to work in build mode).
- It uses "global" `tsconfig.json` which is not used for actual build, but used for jest. This allows to run one instance of jest to execute all tests across packages (including watch mode and coverage mode).

This combination allows running tests or dev/debug build in watch mode for easier development and is producing loosely coupled `dist` folders for distribution.

evgeniuz commented:
Fixed docker build, but not sure what's with circleCI.

evgeniuz commented:
Technically everything should be complete, but CircleCI tests are timing out for some reason (work locally and don't take too much time). Cannot troubleshoot this as output is truncated.

evgeniuz commented:
Cause for failure was that it was requesting too many workers and correspondingly allocating too much memory. Fixed by limiting workers to 2.

evgeniuz commented on packages/ens/src/index.ts:
```diff
@@ -1,7 +1,10 @@
+import { EthqlPluginFactory } from '@ethql/base/dist/plugin';
 import resolvers from './resolvers';
 
+import {} from '@ethql/base/dist/core/services/web3';
```
I think it’s correct way to approach this: if you intend to publish all packages on NPM, there will be no TS sources in packages, only compiled files. So if we refer to TS files, these imports will be unresolved when installed from NPM.

No types information is lost, as all those compiled files always have accompanying .d.ts with them.

Ideally, base package should export single index.ts file that will contain all other exports in it to remove paths and just import @ethql/base everywhere, but I think this should be done after this PR is integrated.

evgeniuz replied:
And tsc build mode intended to work with this approach: it checks if referenced compiled import is present and if not present it builds it from TS.

evgeniuz commented on packages/ens/src/resolvers/index.ts:
```diff
@@ -1,6 +1,6 @@
 import * as _ from 'lodash';
-import { EthqlContext } from '../../context';
-import { EthqlBlock } from '../../core/model';
+import { EthqlContext } from '@ethql/base/dist/context';
```
Please, see above for comment on this.

evgeniuz commented on packages/base/src/services.ts:
```diff
@@ -8,6 +8,8 @@ export type EthqlServiceDefinition<Config, Service> = {
   };
 };
 
+export interface EthqlServiceDefinitions {}
```
This interface is extended in every service but had no "empty" definition. This caused problem compiling `@ethql/base` as there is just one service and it's not imported anywhere in itself (it's imported in `ethql` package, but since we need to be able to build `@ethql/base` in isolation, this is not relevant here)/

raulk replied:
👍

evgeniuz commented:
Are there any changes I need to implement here?

raulk commented:
I pushed a commit a few days ago performing some further refactor, but didn't have time to comment on it. Here it goes:

* Teased out the core plugin from the base module. This was pretty traumatic because it required quite a bit of disentanglement, but the result is much cleaner.
* Extracted the web3 typings to a module of its own.
* Added `main` and `typings` fields in `package.json` files, and added `index.ts` files where they were missing, making importing more idiomatic.

raulk commented:
Just pushed a few more commits to rationalise the build commands; hopefully it'll now pass on CI. I also fixed the Dockerfile.
